certbot/compat/filesystem.py

```diff
@@ -36,6 +44,60 @@ def chmod(file_path, mode):
         _apply_win_mode(file_path, mode)
 
 
+def open(file_path, flags, mode=0o777):  # pylint: disable=redefined-builtin
+    # type: (str, int, int) -> int
+    """
+    Wrapper of original os.open function, that will ensure on Windows that given mode
+    is correctly applied.
+    :param str file_path: The file path to open
+    :param int flags: Flags to apply on file while opened
+    :param int mode: POSIX mode to apply on file when opened,
+        Python defaults will be applied if ``None``
+    :returns: the file descriptor to the opened file
+    :rtype: int
+    """
+    if POSIX_MODE:
+        # On Linux, invoke os.open directly.
+        return os.open(file_path, flags, mode)
+
+    # Windows: handle creation of the file atomically with proper permissions.
+    if flags & os.O_CREAT:
+        # If os.O_EXCL is set, we will use the "CREATE_NEW", that will raise an exception if
+        # file exists, matching the API contract of this bit flag. Otherwise, we use
+        # "CREATE_ALWAYS" that will always create the file whether it exists or not.
+        disposition = win32con.CREATE_NEW if flags & os.O_EXCL else win32con.CREATE_ALWAYS
+
+        attributes = win32security.SECURITY_ATTRIBUTES()
+        security = attributes.SECURITY_DESCRIPTOR
+        user = _get_current_user()
+        dacl = _generate_dacl(user, mode)
+        security.SetSecurityDescriptorDacl(1, dacl, 0)
+
+        try:
+            handle = win32file.CreateFile(file_path, win32file.GENERIC_READ,
```

Comment:
just to check -- does this behave the same way as it would on linux if the file exists, but is a directory?

Reply:
No, in this case you have a `PermissionError` while in Linux it is a `FileExists` (`EEXIST`), since on Windows, files and directories are two different objects.

However the behavior of `filesystem.open` is the same than standard `os.open` on Windows in this situation.

Reply:
as in, we'll hit the PermissionError here that we otherwise would have hit later, in the call to `os.open`?

Reply:
Could you be more precise about what you have in mind here, in order to answer more accurately?

Indeed the PermissionError occurs in CreateFile call instead of os.open, and so ensures to have the same behavior and the same result than the original os.open method.

Reply:
Ok, so that does mean we should be catching PermissionError, and raising it as EEXIST instead? Is there a list of errors CreateFile might raise?

Reply:
You want to make `filesystem.open` not behave like `os.open` on Windows? What is the reason?

Reply:
Also about your question for the list of errors. No, there is not. We are on Windows here, not some sort of open source software with people willing to write exhaustive and comprehensive documentation ^^

Reply:
Isn't the whole point of this to get consistent behavior out of `filesystem.open` across platforms?

Reply:
Really good question!

In fact it is not exactly this goal. The point is to make Certbot behaves correctly, and this imply most of the time to get consistent behavior across platforms. But not always, for instance the requirement to run Certbot as an administrator, while you can run Certbot as a normal user on Linux. 

It means that I do make converge behaviors across platforms not by principle, but by necessity.

I prefer to avoid converting this error into a `EEXIST` for the following reasons:
* Certbot works correctly without it
* Python developers did not convert this windows exception into a python one, which constitute a hint here in my opinion
* and in fact, the error is very generic: it is a `PermissionError`, not something like "a directory already exists at this path"

This makes me worry to accidentally shadow an unrelated exception into an EEXIST one, and create unexpected side effects in the future. Also as you said, it is not documented. It seems that no Windows developer thought it would be a good idea to document, and maybe handle specifically this case on a so much critical function as `CreatFile` is.

Instead you just get a "You are not authorized to do the action you are trying to do"...

Reply:
Ah, I see now. The situation I was imagining is that somewhere else in certbot, we wrap the function call that was previously `os.open` and is now `filesystem.open` in a `try`, and catch `EEXIST` only, instead of setting flags in the method call. In which case, I was worried that we'd crash there when really we should have caught it in the hypothetical `except`.

But since `PermissionError` is more generic, we probably do want to be letting it raise, since it could mean other things as well.

I still feel a little weird about leaving it to stumble across in the future... at minimum, documenting with a `:raise:`would be helpful, open to other suggestions as well.

Reply:
Yes, this behavior is really unfortunate on Windows. But the segregation between files and directories is at the core design of the Windows filesystem.

To enlight this a little more, I tried to invoke `win32file.CreateFile` using the `win32con.CREATE_ALWAYS` flag on a situation where a directory of the same name than the file that we try to create. To recall, `win32con.CREATE_ALWAYS` instruct Windows to always re-create the file, even if it already exists. But if it is a directory, you end again with a FilePermission error.

Reply:
and on Linux you get an `OSError`. Looks like `PermissionError` subclasses `OSError`. So the most specific thing we can promise is that we'll throw an `OSError`, which is fine imo. But we should note that in the docstring.

Reply:
I added a `:raise:` entry in the docstring.